docs: Auth-Layer design doc (#32)
hashgraph/hedera-the-graph: docs/design/auth-layer.md

We will be using the following endpoints of the Keycloak server:
- **Token:** `/realms/HederaTheGraph/protocol/openid-connect/token` to obtain the access token.
- **Token Introspect:** `/realms/HederaTheGraph/protocol/openid-connect/token/introspect` to validate the access token and retrieve token claims.

Users permissions will be defined with 3 custom roles: "subgraph_create", "subgraph_deploy" and "subgraph_remove". And with a custom claim called "subgraph_access" that will contain a CSV String with the subgraphs names that the user has access to.

#### Roles: 
1. **subgraph_create:** allows the user to create a new subgraph.
2. **subgraph_deploy:** allows the user to deploy a subgraph.
3. **subgraph_remove:** allows the user to remove a subgraph.

#### Custom Claims:
- **subgraph_access:** contains a CSV String with the subgraphs names that the user has access to. ie: "subgraph1,subgraph2,subgraph3"

#### Token Request example:

```bash

> curl --location 'http://host.docker.internal:8080/realms/HederaTheGraph/protocol/openid-connect/token' \

Reviewer: There's a few fields in this call that need to exist first.
You should call out what expected setup has already performed e.g. a user profile is already setup with x y and Z configs

Author: ✅ Done

>     "expires_in": 36000,

Reviewer: Can expiration be configured and if so by who and where?

Author: Yes, it can be configured by the node administrator on keycloak, and most likely it should be a lot higher number. to avoid requesting for a new token too often.

    "typ": "Bearer",
    "azp": "htg-auth-layer",
    "session_state": "aa53330d-ff69-4ea5-8b4f-7d49c964740d",
    "acr": "1",
    "allowed-origins": [
        "/*"
    ],
    "realm_access": {
        "roles": [
            "default-roles-hederathegraph",
            "offline_access",
            "uma_authorization"
        ]
    },
    "resource_access": {
        "htg-auth-layer": {
            "roles": [
                "subgraph_create",
                "subgraph_deploy"
            ]
        },
        "account": {
            "roles": [
                "manage-account",
                "manage-account-links",
                "view-profile"
            ]
        }
    },
    "scope": "email profile subgraph_access",
    "sid": "aa53330d-ff69-4ea5-8b4f-7d49c964740d",
    "email_verified": false,
    "name": "Developer Example",
    "subgraph_access": "subgraph1,subgraph2,subgraph3",
    "preferred_username": "developer1",
    "given_name": "Developer",
    "family_name": "Example",
    "email": "developer1@gmail.com",
    "client_id": "htg-auth-layer",
    "username": "developer1",
    "token_type": "Bearer",
    "active": true
}
```

### D. Hedera-TheGraph Node

> This component is responsible for indexing data from the Hedera network and storing it in a database. The Hedera-TheGraph node is also responsible for serving queries from the GraphQL API. The Hedera-TheGraph node will be modified to include the authentication proxy and to validate the access token before processing requests.

Reviewer: How will it be modified?
Should call this out as it suggests this node may differ from the default node

Author: This was a confucion, the node server does not need to be modified to support auth, hence the proxy on top. 😅 
Thanks for noticing.

✅ Already corrected the text

> The Admin API of the Hedera-TheGraph node will be reachable only via the `Auth-Proxy` and will require the access token to be included in the header of the requests, and the `Auth-Proxy` will validate the access token and the user's permissions before forwarding the request to the Hedera-TheGraph node.

Reviewer: Are there any non subgraph create, deploy or remove endpoints that need to be protected or will not be protected on the admin API?

Author: Good Catch, there are other Admin endpoints.

`subgraph_reassign`
`subgraph_pause`
`subgraph_resume`

https://github.com/graphprotocol/graph-node/blob/591ad93b5144ff5e6037b73862c607effad90e7f/server/json-rpc/src/lib.rs#L47-L82

✅  I've added those as roles too (except for subgraph_reassign), so they can be exposed and protected as well, however in practice it might be simpler to create a `Group` that has all the roles, and assign devs those roles, unless more granularity is requested or needed for a specific subgraph project to a member of their team.

### C. Authentication Server

This component is responsible for authenticating users and providing access tokens. The authentication server is responsible for validating the user's credentials and generating an access token. The access token is used to authenticate the user when making requests to the Hedera-The-Graph node. The authentication server will use OAuth2 and JWT standards to provide access tokens, and it will manage the custom claims to define the user's permissions. 

Any authentication server that supports OAuth2 and JWT standards can be used, but for the implementation included on this project we will use a `Keycloak` server, it was chosen because it is a robust OSS Identity Provider that is widely used and has a lot of documentation and community support available and it is easy to integrate with other systems and to manage users, roles, and permissions.

We will be using the following endpoints of the Keycloak server:
- **Token:** `/realms/HederaTheGraph/protocol/openid-connect/token` to obtain the access token.
- **Token Introspect:** `/realms/HederaTheGraph/protocol/openid-connect/token/introspect` to validate the access token and retrieve token claims.

> Users permissions will be defined with 3 custom roles: "subgraph_create", "subgraph_deploy" and "subgraph_remove". And with a custom claim called "subgraph_access" that will contain a CSV String with the subgraphs names that the user has access to.

Author: ✅  Done!

#### Roles: 
1. **subgraph_create:** allows the user to create a new subgraph.
2. **subgraph_deploy:** allows the user to deploy a subgraph.
3. **subgraph_remove:** allows the user to remove a subgraph.
4. **subgraph_pause:** allows the user to pause indexing a subgraph.
5. **subgraph_resume:** allows the user to resume indexing a subgraph.

Each role will enable the user to consume a protected endpoint on the Hedera-The-Graph node's Admin API. There is a 1:1 relationship between the roles and the protected endpoints. 

*The only endpoint that is not supported is `subgraph_reassign` because its use is not recommended and it is not a common operation, besides it does not make sense to reassign a subgraph to another indexer, because by default Hedera-The-Graph node is a single indexer setup.

#### Custom Claims:
- **subgraph_access:** contains a CSV String with the subgraphs names that the user has access to. ie: "subgraph1,subgraph2,subgraph3"

#### Token Request example:

> On the example below, the `user` needs to have its email confirmed (activated) and set its own `password`. The `client_id` and the `client_secret` are needed to identify the application within the Keycloak server and would be provided to the developers by the Hedera-The-Graph node administrators.

Reviewer: ```suggestion
On the example below, the `user` needs to have their email confirmed (activated) and `password` set prior to requesting a token. The `client_id` and the `client_secret` are needed to identify the application within the Keycloak server and would be provided to the developers by the Hedera-The-Graph node administrators.
```

Author: I like it, thanks! 👍 

✅  Done!

> ```bash

Author: Yes you are right, the specification is for a keycloak implementation of the auth-server, so I guess it makes sense.

However a note indicating that the flows to obtain a token might change depending the auth-server choosed.

✅  Done!
